[READY] Warm hugs for everyone

nightred opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

If you ever feel down and out, a nice warm hug will brighten your day.
When you hug someone you warm the lowest body temp is warmed up. This accounts for insulation, max temperature change rates, and change steps.

Further changes
* proc natural_bodytemperature_stabilization is now self contained,
* proc adjust_bodytemperature has been expanded for carbon mobs to allow the use of insulation and change steps
* most of handle_environment has been moved to natural_bodytemperature_stabilization and adjust_bodytemperature cleaning it up.
* When hugging both parties temperature is adjusted one up one down, you can not freeze someone without also freezing yourself. 

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

Some time you need a warm hug to get you through the day.
Save someone freezing to death by hugging them.
Ethereal's are nice and warm think of them when you look for that next hug.
Code cleanup round two on handle environment, separating parts out of the old large function

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Hugs are warm now
refactor: natural_bodytemperature_stabilization is now self contained.
refactor: adjust_bodytemperature to allow the use of insulation and change steps
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

nightred committed:
Warm hugs for everyone

HomocidlePancakes commented:
Hugs reducing stamcrit recovery time when?

GuyonBroadway commented:
What if someone is running a feaver from a virus, could you in theory heat people beyond standard?

HomocidlePancakes commented:
Break the laws of thermodynamics with hugs.

XDTM commented:
Xenobiologist can now spacewalk while being hug-warmed by their gang of golems

zxaber commented:
Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

cacogen commented:
Should be able to throw eggs at hot borgs to make fried eggs.

nightred commented:
Borg's are not included.

nightred committed:
To hot to hug time to cool down

nightred commented:
> What if someone is running a feaver from a virus, could you in theory heat people beyond standard?

You can warm people beyond normal body temp, but if you are over the heat damage limit (or they are) you do not want to hug them. This will prevent burns.

XDTM commented:
I wouldn't restrict hugging by temperature, maybe simply add a message when it happens so people know what made them overheat in case they hug a recently-burning person.

Jalleo reviewed (commented):
I hope this doesn't work through rigsuits , firesuits , (bio/rad/bomb)suits , spacesuits and more

nightred committed:
Warm hug mood, and notice on hot hugs

nightred commented:
> I wouldn't restrict hugging by temperature, maybe simply add a message when it happens so people know what made them overheat in case they hug a recently-burning person.

I removed the block on the hug and changed it to a output message.
I also added a new mood when you are warmed up by a hug.

> I hope this doesn't work through rigsuits , firesuits , (bio/rad/bomb)suits , spacesuits and more

Because this checks for insulation you are not warmed up when wearing full insulation, and the more insulation you have the worse it works.

Mickyan commented on code/datums/mood_events/generic_positive_events.dm:
```diff
@@ -19,6 +19,14 @@
+	timeout = 2 MINUTES
+
+/datum/mood_event/warmhug/add_effects()
+	description = "<span class='nicegreen'>Someone gave me a warm cozy hug!</span>\n"
```
what's this for?

nightred replied:
When you click the moodlet icon it reports this back to you. Am I wrong?

Mickyan replied:
that's what the description in **/datum/mood_event/warmhug** does

You only need this proc if you, say, want the message to include the name of the person that hugged you, like in **/datum/mood_event/besthug**

nightred committed:
fixed mood, added insulation proc

ATH1909 commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,28 @@
+
+		// Let people know if they hugged someone really warm
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>[M] felt like they are over heating when they hugged you.</span>")
```
tenses

nightred replied:
think i fixed it up.

ATH1909 commented:
This should cool down/heat up the person performing the hug as well. Something something laws of thermodynamics.

Can you cool yourself down to near 0K as a plasmaman, then hug lizardpeople to kill them with cold?

ATH1909 commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,28 @@
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>[M] felt like they are over heating when they hugged you.</span>")
+
+		if(bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
```
add messages for hugging/being hugged by people who are freezing too

nightred replied:
done

ATH1909 commented:
> Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

I doubt it; their hugging modules don't even help people recover from being stunned or knocked unconscious faster.

nightred committed:
Balance heat exchange, notify when freezing also

nightred commented:
> This should cool down/heat up the person performing the hug as well. Something something laws of thermodynamics.

You are now cooled when you hug someone. Insulation and difference range, and max heating cooling steps are still accounted for.

> Can you cool yourself down to near 0K as a plasmaman, then hug lizardpeople to kill them with cold damage?

Sort of, the warm person will cool down, the cold person will warm up. At 0 kelvin a human at 310 kelvin will heat you up about 20 degrees k per hug, and you will cool down about that much also.

> Does this work for borgs? I look forward to a borg sitting for a few minutes in a fire and then hugging people to burn them.

This only work on mob/living/carbon/ sub types, borgs are part of mob/living/silicon and can not directly interact with carbons the same way.

nightred committed:
tenses again

ShizCalev commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,40 @@
+
+		// Let people know if they hugged someone really warm or really cold
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>It feels like [M] is over heating when they hug you.</span>")
```
overheating as they hug you

nightred replied:
Corrected thanks

ShizCalev commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,40 @@
+		if(M.bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>It feels like [M] is over heating when they hug you.</span>")
+		else if(M.bodytemperature < BODYTEMP_COLD_DAMAGE_LIMIT)
+			to_chat(src, "<span class='warning'>It feels like [M] is freezing when they hug you.</span>")
```
as

nightred replied:
Correcting also.

ShizCalev commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,40 @@
+			to_chat(src, "<span class='warning'>It feels like [M] is freezing when they hug you.</span>")
+
+		if(bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(M, "<span class='warning'>It feels like [src] is over heating when you hug them.</span>")
```
.

ShizCalev commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,40 @@
+		if(bodytemperature > BODYTEMP_HEAT_DAMAGE_LIMIT)
+			to_chat(M, "<span class='warning'>It feels like [src] is over heating when you hug them.</span>")
+		else if(bodytemperature < BODYTEMP_COLD_DAMAGE_LIMIT)
+			to_chat(M, "<span class='warning'>It feels like [src] is freezing when you hug them.</span>")
```
.

ShizCalev commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -631,6 +631,20 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
+ * return the percentage of protection as a value from 0 - 1
+**/
+/mob/living/carbon/proc/get_insulation_protection(temperature)
+	if (temperature > bodytemperature) // exposed to heat, get heat protection
```
condense this as so

return (temperature > bodytemperature) ? get_heat_protection(temperature) : get_cold_protection(temperature)

nightred replied:
Changed also

nightred committed:
words, and proc clean up

Cyberboss commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,40 @@
-		SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug)
+
+		// Warm them up with hugs
+		if(bodytemperature > M.bodytemperature) // they are warmer leech from them
```
This feels like copypasta from the temperature PR. If I'm right, can we make them share the code somehow please?

nightred replied:
This is not a direct copy, but sharing body heat might be good to get as a proc somewhere in the mob framework. I have an idea let me see.

nightred replied:
Ok Cyber I did some major changes.
natural_bodytemperature_stabilization is now self contained.
adjust_bodytemperature is expanded for carbon mobs to allow insulation and change steps
handle_environment has been reduced in size as most of it is now done in the other two steps.

nightred committed:
Common interfaces

nightred committed:
fixed monkey override

ExcessiveUseOfCobblestone commented:
they should share heat instead of getting "free" heat.

nightred commented:
> they should share heat instead of getting "free" heat.

They do share heat now actually, in the code that is. i guess i did not document that, fixing.

nightred committed:
lets not have another hard suit situation

nightred commented:
Issue still with hard suits

GuyonBroadway commented:
Idea for a followup pr.

Hugs spread radiation for extra spicy hugging.

nightred commented:
> Idea for a followup pr.
> Hugs spread radiation for extra spicy hugging.

Funny thought but just being near someone with rads will give you rads so no need to do anything extra. Do not hug glowers.

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -679,15 +679,22 @@ GLOBAL_LIST_INIT(ballmer_windows_me_msg, list("Yo man, what if, we like, uh, put
 	if(use_steps)
-		amount = (amount > 0) ? min(amount / BODYTEMP_HEAT_DIVISOR, BODYTEMP_HEATING_MAX) : max(amount / BODYTEMP_COLD_DIVISOR, BODYTEMP_COOLING_MAX)
+		amount = (amount > 0) ? min(amount / BODYTEMP_HEAT_DIVISOR, BODYTEMP_HEATING_MAX) : \
+			max(amount / BODYTEMP_COLD_DIVISOR, BODYTEMP_COOLING_MAX)
```
1 line please

nightred replied:
fixed this up to be wide style

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1633,7 +1633,8 @@ GLOBAL_LIST_EMPTY(roundstart_races)
 	if(!H.on_fire || areatemp > H.bodytemperature) // If we are not on fire or the area is hotter
-		H.adjust_bodytemperature((areatemp - H.bodytemperature), use_insulation=TRUE, use_steps=TRUE)
+		H.adjust_bodytemperature((areatemp - H.bodytemperature), use_insulation=TRUE, use_steps=TRUE, \
+			hardsuit_fix=bodytemp_normal - H.bodytemperature)
```
same here

nightred replied:
fixed this up to be wide style

nightred committed:
really wide lines

nightred committed:
need to invert that number

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/carbon_defense.dm:
```diff
@@ -267,7 +267,26 @@
+			SEND_SIGNAL(M, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, src) // Hugger got a warm hug
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/hug) // Reciver always gets a mood for being hugged
+		else
+			SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "hug", /datum/mood_event/warmhug, M) // You got a warm hug
```
why don't you just unindent

nightred replied:
So that I do not overwrite a mood set one before, and only the cold hugger gets the warm mood.

ExcessiveUseOfCobblestone replied:
I literally saw this and don't know why I still sent the comment cringe

ExcessiveUseOfCobblestone commented on code/modules/mob/living/carbon/human/species.dm:
```diff
@@ -1782,31 +1748,61 @@ GLOBAL_LIST_EMPTY(roundstart_races)
+ * * environment The environment gas mix
+ * * H The mob we will stabilize
+ */
+/datum/species/proc/natural_bodytemperature_stabilization(datum/gas_mixture/environment, mob/living/carbon/human/H)
```
is this a duplicate of the carbon still? they look awfully similar

nightred replied:
This version uses the species body temperature customizations.

ExcessiveUseOfCobblestone replied:
rad
